### PR TITLE
misc: refactor plugin

### PR DESCRIPTION
**Pull request by jkroepke**
Passing all request contexts to chrome instances
Enable plugin logger for chromedp context

I didn't test that. But on theory, if the http server shutdown gracefully, all requests are canceled first. In conclusion, any chrome contexts should be canceled as well.

### CODE REVIEW

**Comment by mahendrapaipuri** [comment]
Thanks again for this feature. I agree that it is more clean passing the request context everywhere. Just a minor comment: if we are passing `context` to Grafana client, shouldnt we make `RequestWithContext` [here](https://github.com/jkroepke/grafana-dashboard-reporter-app/blob/e124b901cbb0d797dd6d0aa060fbf2e4a5c88bb2/pkg/plugin/client.go#L161)?

Cheers for setting up logger for `chromedp` as well. That can provide more useful logs for users.

**Comment by jkroepke** [comment]
> shouldnt we make `RequestWithContext`

Highly recommend to introduce [golang-ci](https://github.com/golangci/golangci-lint), because it would trigger a warning for this case, since it embeds https://github.com/kkHAIKE/contextcheck and https://github.com/sonatard/noctx

**Comment by mahendrapaipuri** [comment]
Agree with `golangci-lint`. I will set that up in a separate PR. However, I have few ideas and suggestions for #51 and if we implement them, this PR most probably will become obselete. So, I suggest we put this on hold and move the discussion to #51 to discuss the feature further.

**Comment by jkroepke** [comment]
the chrome logger and context is still worth.

I would consider to merge it and if some are getting reverted as part of #51 - it is fine.

Otherwise, please close the PR.

**Comment by mahendrapaipuri** [comment]
> the chrome logger and context is still worth.

Yes, I agree. But if we manage the lifecycle of chromium from [`NewApp`](https://github.com/jkroepke/grafana-dashboard-reporter-app/blob/44e6305583d7006184cfc98a855e17d436fde667/pkg/plugin/app.go#L150) function, we will setup chromium logger from context logger in `NewApp` and so as the context. That will make `client.go` and `report.go` way more simpler as they will simply the consumers of configured chromium instance without having to do any new configuration on chromium. 

Globally what I am suggesting is move all the configuration to `NewApp` which is supposed to manage the plugin instance and make downstream plugin consumer of configured resources. Does it make sense?

**Comment by jkroepke** [comment]
Hi, feel free to commit your suggestions. 

I do not have the full overview of the current architecture of the plugin. But I agree that some interfaces feels a bit complicated.

**Comment by jkroepke** [comment]
Over the last weeks, by on my Golang experience, I do some refactoring of the plugin. I followed the recommendation todo one big PR:

Summarize of all changes:

- separate code into go packages
- Remove global config state in App
- most of the chrome related logic is inside separate browser instance client now
- graceful shutdown for chrome
- get error from chrome, if there are errors opening grafana dashboard
- Add remote chrome instances with tests
- PDF files will be streamed from chrome to client. This avoid that the whole PDF is stored in memory
- Panel PNGs files will be streamed through base64 encoder. This avoid that the whole PNGs are stored twice in memory
- max render worker are global now. This avoid high resource usage on concurrent requests
- added browser worker. This avoid high resource usage on concurrent requests.
- url is now overridable via plugin config. If url is unset, the app url from Grafana will be used.
- GET parameter will be validated now.
- Rename skipTlsCheck to tlsSkipVerify, because its Grafana standard. ([ref](https://github.com/grafana/grafana-plugin-sdk-go/blob/c5aa822be58253fd954e921aec00241833d9ca4c/backend/http_settings.go#L221))
- Instead using GF_APP_URL, the backend SDK is used to gain the appURL. The appURL can be overwritten to avoid additional hops. The environment variable is [deprecated](https://github.com/grafana/grafana-plugin-sdk-go/blob/c5aa822be58253fd954e921aec00241833d9ca4c/backend/config.go#L155) by Grafana.

**Comment by mahendrapaipuri** [comment]
Thanks a lot for this awesome work @jkroepke 

I will need some time to look into it as it is quite a big one. But it looks very promising and I like the refactoring you did. I think it would be great addition to the plugin. Thanks again for all your time and effort. Really appreciate it.

File: pkg/plugin/internal/dashboard/errors.go
Comment (mahendrapaipuri) [inline pkg/plugin/internal/dashboard/errors.go:1]: I dont really see the value of this file. Either we define all the custom errors used in the package here or remove this file and use `fmt.Errorf` as we are doing everywhere.

File: pkg/plugin/internal/client/errors.go
Comment (mahendrapaipuri) [inline pkg/plugin/internal/client/errors.go:1]: Same here as well. Either define all custom errors here or remove this file and use `fmt.Errorf`

**Review by mahendrapaipuri — CHANGES_REQUESTED** [review]
I have few minor comments. I really like the new sub packages which is more clear and organised. The only major change is I would prefer all the packages **taken out** of `internal`. Something like [grafana-infinity-datasource](https://github.com/grafana/grafana-infinity-datasource/tree/main/pkg) organization. I feel that would be more intuitive for developers to follow the code.

I think we should expand e2e test matrix and update docs. We can do it in next step in a separate PR.

File: .editorconfig
Comment (jkroepke) [reply .editorconfig:1]: tbh, fine for CI, but its makes it easier for local editors. but I can remove it.

File: pkg/plugin/app.go
Comment (jkroepke) [reply pkg/plugin/app.go:69]: This is bit problematic, since skip TLS will not handle by `config.Load`. 

The skip TLS config is handled by Grafana SDK plugin it-self (https://github.com/grafana/grafana-plugin-sdk-go/blob/c8a47594ccdb281c69df4fbd74f7f233c64ab25b/backend/http_settings.go#L222), this env var is more or less a hack and i would completely remove it. That would make it easier

File: pkg/plugin/internal/config/settings.go
Comment (jkroepke) [reply pkg/plugin/internal/config/settings.go:15]: `tlsSkipVerify` is standarized names from the Grafana SDK. That was the reason for the rename. By using `tlsSkipVerify`, the Grafana SDK can configure the httpClient out of the box.

https://github.com/grafana/grafana-plugin-sdk-go/blob/c8a47594ccdb281c69df4fbd74f7f233c64ab25b/backend/http_settings.go#L222

app is a standardized name, too. But it has not affect on apps. I can rename it back.

File: pkg/plugin/internal/dashboard/errors.go
Comment (jkroepke) [reply pkg/plugin/internal/dashboard/errors.go:1]: Errors, produced inside a package, should be defined as package level variable inside the package. That part of a well designed error handing which exists since Go 1.13

https://go.dev/blog/go1.13-errors#errors-and-package-apis

There is also a linter for that, included in golangci-lint: https://github.com/Djarvur/go-err113

File: pkg/plugin/internal/report/templates/report.gohtml
Comment (jkroepke) [reply pkg/plugin/internal/report/templates/report.gohtml:79]: But, currently, we do not request PNG. The embed function require the original image , but EncodedImage already contains the base64 encoded image. 

> We can request JPEG panels

I see no reason for make this flexible, if PNGs are always hard-coded anyways on request. But I may have an idea to make it flexible.

File: provisioning/plugins/app.yaml
Comment (jkroepke) [reply provisioning/plugins/app.yaml:41]: please check first comment about that.

File: pkg/plugin/internal/config/settings.go
Comment (jkroepke) [reply pkg/plugin/internal/config/settings.go:28]: Why? Could you explain the benefit?

File: pkg/plugin/app.go
Comment (jkroepke) [reply pkg/plugin/app.go]: I was every asking about, if an admin changes settings on the Admin UI, how settings are propagate to the plugin?

It seems like there was an issue in the SDK. Ref: https://github.com/grafana/grafana-plugin-sdk-go/pull/1039

Comment (mahendrapaipuri) [reply pkg/plugin/app.go]: Well, everything the settings are changed in the Grafana UI, the plugin app will get reloaded. This mean the current plugin instance will be disposed and a new instance with updated config will be created. So, the subsequent requests after the reload will see the _new updated_ config. That is what I understood from reading the code and doing some tests. The only thing is they wait for like 5 seconds or so to dispose the existing plugin instance presumbaly to finish any requests in progress.

File: pkg/plugin/internal/config/settings.go
Comment (mahendrapaipuri) [reply pkg/plugin/internal/config/settings.go]: I think I havent explained myself well. Sorry about that. I meant to ask to change the JSON tags to `appUrl` and `skipTlsCheck` as they were before. There is no big impact on changing the field names on package side and if we can configure HTTP client out-of-the-box with new names, that would be fine with me as well.

I just dont want to change the user facing config parameter names as they will break the existing deployments when we cut the next release. I really dont fancy spending a lot of time answering people's issues on why they are broken. In my experience a lot of people do not look at changelogs and docs. Does it make sense?

But I see now you removed the `skipTlsCheck` from the config struct and that **was not my intention**. We should still keep it and let operators to provision this parameter as well. Ideally I would like to give an option to operators to configure the app either by provisioning or by env vars or using combination of both. That is what both Grafana and grafana-image-renderer does. My idea was to use [go-envconfig](https://github.com/sethvargo/go-envconfig) to be able to set any struct field with an env var as well **besides** provisioning. Its fine, I can make a separate PR for that once we merge this.

File: pkg/plugin/internal/report/templates/report.gohtml
Comment (mahendrapaipuri) [reply pkg/plugin/internal/report/templates/report.gohtml]: Well, the response we get from Grafana when fetching panel images will not add headers automatically. That is why `embed` function was there to detect the format and add headers. Infact the users can pass query parameters to report API that are specific to `grafana-image-renderer` and they will be passed through to renderer. So users can control the format, if they need and by using `embed` function we dont need to do anything extra to account for these different mime formats.

File: pkg/plugin/app.go
Comment (jkroepke) [reply pkg/plugin/app.go]: Thats what I observe, too. Check latest commits.

File: pkg/plugin/internal/dashboard/errors.go
Comment (mahendrapaipuri) [reply pkg/plugin/internal/dashboard/errors.go]: I understand that design pattern for errors that are reused in the package. But the errors you defined is only used once. Then the question is why defining just these errors? There are plenty of custom `fmt.Errorf()` in package and we should move all of them into `errors.go`. I guess the linter suggests to replace `fmt.Errorf()` with `errors.New()`. But I prefer to keep using `fmt.Errorf()` formating the error is straight forward where as we need to use `fmt.Sprintf()` inside `errors.New()`.

As each function returns its own custom error, I think its better to keep it close to the function. If we are reusing the same error atleast twice then it justifies definining it in new `errors.go`. Does it make sense?

**Review by mahendrapaipuri — CHANGES_REQUESTED** [review]
Please fix the failing CI jobs and I can push the rest of little changes that we need before merging this PR.

Thanks again for this awesome work. You put a lot of effort and time and I really appreciate it. You really improved the quality of code a lot with your insights.

File: pkg/plugin/internal/dashboard/errors.go
Comment (jkroepke) [reply pkg/plugin/internal/dashboard/errors.go]: > Then the question is why defining just these errors

That current error handling design in the go language. Error, that coming from that packages, should be created as package level initiated error. Error, coming from external packages, should be wrapped with `fmt.Errorf` together with the `%w` placeholder.

The goal is that caller can be used `errors.Is` to check for an specific error. Before go 1.13, string compare where is to indicate specific errors. In tests, I could also check for specific errors.

https://github.com/jkroepke/grafana-dashboard-reporter-app/blob/a7a685612010902d854c6da7157caa06d891143b/pkg/plugin/client/client_test.go#L89

Basic each guide to error handling leads to that current implementation.

> There are plenty of custom fmt.Errorf()

This is not technically correct. I go through most of the error and we are not creating any new custom errors. Instead, with `fmt.Errorf`, we wrap returned errors from external packages. The different is `%w` on the errors, we can't be inside an errors.New declaration. 

Please do not mix creating/define new errors and wrap errors.

> I think its better to keep it close to the function.

That is still the case:

for example:

```go
	if len(dashboardData) == 0 {
		return nil, ErrJavaScriptReturnedNoData
	}
```

It's visible, which error is returned. It required to give package level errors verbose names.

---

What may could help to understand that concept:

with `var ErrJavaScriptReturnedNoData = errors.New("javascript did not return any dashboard data")`, it's the go-away to defined new Exception Types and in Java or JavaScript, you may also define multiple custom exceptions for different scenarios. While in java `try {} catch {ErrJavaScriptReturnedNoData e}` exists, in go you just using `if errors.Is(err, ErrJavaScriptReturnedNoData)`

---

Additional resource:

- https://tech.new-work.se/designing-errors-with-go-1-13-81692537457
- https://go.dev/blog/go1.13-errors

Example for go stdlib:

- https://github.com/golang/go/blob/1f0c044d60211e435dc58844127544dd3ecb6a41/src/flag/flag.go#L99-L109
- https://github.com/golang/go/blob/1f0c044d60211e435dc58844127544dd3ecb6a41/src/net/interface.go#L31-L37
- https://pkg.go.dev/io#ErrClosedPipe

File: pkg/plugin/resources.go
Comment (jkroepke) [reply pkg/plugin/resources.go:101]: good catch

Comment (jkroepke) [reply pkg/plugin/resources.go]: It resolves the CI failures, too

**Comment by jkroepke** [comment]
CI is green now, feel free to finish and feel free revert the error handling, if you are not agree with that.

File: pkg/plugin/internal/dashboard/errors.go
Comment (mahendrapaipuri) [reply pkg/plugin/internal/dashboard/errors.go]: Ahhh, I was not thinking about `errors.Is()`. Yes, I agree with you in that case, they deserve to be declared. Thanks for the detailed explanation. Appreciate it!

**Comment by mahendrapaipuri** [comment]
> CI is green now, feel free to finish and feel free revert the error handling, if you are not agree with that.

I think we are good. There are very small changes that I will do them in a separate PR. 

I dont want to squash merge this PR as we will loose a lot of context doing so. Do you agree squashing the commits with same commit message? If you agree, could you please do so and force push the branch? If not I can do it just let me know.

Thanks again for the work. Codebase looks much better now.

**Comment by jkroepke** [comment]
Is this fine now?

**Review by mahendrapaipuri — APPROVED** [review]
Amazing work!! Cheers again!

**Comment by jkroepke** [comment]
@mahendrapaipuri Any plans for a release? I'm happy to start the integration in our eco system.

I'm aware that there are small changes todo, but I would appreciate an release candidate, too.

**Comment by mahendrapaipuri** [comment]
@jkroepke I had a deadline, so, couldnt come back to it. I have put up a PR with final changes and some fixes. Feel free to take a look at it. I will try to do a release if everything goes green by end of the day.

**Comment by mahendrapaipuri** [comment]
@jkroepke New release [v1.5.0](https://github.com/mahendrapaipuri/grafana-dashboard-reporter-app/releases/tag/v1.5.0) has been published!

**Comment by jkroepke** [comment]
Thanks alot!

**Pull request merged**
